Improve: add ancestors functions

SlySven opened the pull request:
#### Brief overview of PR changes/additions and motivation for adding

Provides two Lua functions:
* `isAncestorsActive(itemID, "type")` - returns a boolean which is `true` if all (if any) of the parents of the given item are set as active and `false` if any are not. It makes use of an internal template function that already exists but makes it available to the Lua sub-system. If the item does not have any ancestors this function will return `true`. 
* `ancestors(itemID, "type")` - returns a table containing a sub-table for each successively distance ancestor (if any) of the given item; within each sub-table are details of the ancestor, specifically:
  * its ID as a number
  * its name as a string
  * whether it is active as a boolean
  * its "node" (type), one of "item", "group" (folder) or "package" (module)

***ALSO: adds a third, optional, boolean argument to `isActive(...)` which if provided and `true` will only includes items in the count whose parents are all enabled/active - if omitted it will be treated as if it is `false` so that it behaves as the current version. Script-writers should check for the presence of one of the other two functions to determine if this third argument is handled.***

#### Other info (issues closed, discussion etc)
The first function will be enough to close #6724, specifically when it is combined with `isActive(...)` as follows, it will enable a script to determine whether an item (in this example a trigger with an ID number stored as `myTriggerID`) will operate or not:
```lua
    if isAncestorsActive(myTriggerID, "trigger") and isActive(myTriggerID, "trigger") > 0 then
       -- Okay the aforesaid trigger will run!
    end
```

Note that these two new functions ONLY work with the ID number of an item - this is because a unique identifier is required and Mudlet allows for duplicate names so that cannot be used to reliably identify an item. At present it is not as easy as it could be to determine the ID number of an existing item; I believe it would be beneficial to also provide a `findItems("name", "type")` function that could provide a list of all the IDs of items of a specified type that match a given name - this is pending as #6742.

SlySven committed:
Improve: add ancestors functions

Provides two Lua functions:

* `isAncestorsActive(itemID, "type") - returns a boolean which is `true` if
all (if any) of the parents of the given item are set as active and `false`
if any are not. It makes use of an internal template function that already
exists but makes it available to the Lua sub-system. If the item does not
have any ancestors this function will return `true`. This will be enough to
close #6724.

* `ancestors(itemID, "type") - returns a table containing a sub-table for
each successively distance ancestor (if any) of the given item; within
each sub-table are details of the ancestor, specifically:
  * its ID as a number
  * its name as a string
  * whether it is active as a boolean
  * its "node" (type), one of "item", "group" (folder) or "package" (module)

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 reviewed (commented):
I'd like to see the original `isActive` function to be fixed - I don't think we'll be breaking backwards compatibility in an adverse way here. At least I don't see a way we will. @demonnic @keneanung ?

demonnic commented:
I agree, this feels more like a bug with isActive than something which requires a new function to me.

SlySven commented:
An item will be disabled if it or any of it's ancestors are not enabled. So, how do you currently find out which one is not active in a script? As I see it, you cannot!  Hell, you cannot currently even find out which other items ***are*** its ancestors. Changing `isActive(...)` to behave as the proposed `isAncestorsActive(...)` won't tell you which ones are the inactive one(s) either.

demonnic commented:
Then perhaps the `ancestors` function could still have value, but I still feel like the isAncestorsActive isn't necessary. What if multiple items have the same name? If we only include items in the count for isActive which are actually able to fire (effectively active, not just set to active itself as an individual item) then it resolves the issue with isActive itself. 

Perhaps this should be two PRs instead, one for a bugfix for isActive, and one to add the ancestors function?

SlySven commented:
> ... What if multiple items have the same name? If we only include items in the count for isActive which are actually able to fire (effectively active, not just set to active itself as an individual item) then it resolves the issue with isActive itself.

But if you did that using `isActive(...)` to see if an (any) item is capable of working and you got a `0` result you do not know if the situation can be resolved by setting the item itself to be active or if you have to find a parent and make that active *instead* or *as well* - this is two separate matters and the resolution of one is not the same as the other.... :man_shrugging:

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions

SlySven committed:
Standarize UI files

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions

vadi2 commented:
I'm having a hard time following this (and I suspect others will too). What about a simpler design where `isActive` considers ansectors by default (bugfix unless it breaks someone's scripts in PTB testing), and we add a boolean parameter to `isActive` not to consider ancestors?

SlySven commented:
> I'm having a hard time following this (and I suspect others will too). What about a simpler design where `isActive` considers ansectors by default (bugfix unless it breaks someone's scripts in PTB testing), and we add a boolean parameter to `isActive` not to consider ancestors?

:thinking:  I could live, I suppose, with an optional (boolean) argument to `isActive(...)` but the default must be to only consider the identified item - otherwise that breaks the **backwards script compatibility promise**. Also, it must be stressed that if this argument is present and applied then the **number** that that function returned represents the total of items that are actually going to operate because both they and their ancestors are ALL "active" and how that is different than if it is not (and how previous Mudlet version prior to this change operated) because *that* (was/)is only checking the state of the identified item(s)...

vadi2 commented:
> 🤔 I could live, I suppose, with an optional (boolean) argument to isActive(...) but the default must be to only consider the identified item - otherwise that breaks the backwards script compatibility promise.

Yeah, I am now on the fence if we are crossing a line between a bugfix and breaking backwards compatibility.

> Also, it must be stressed that if this argument is present and applied then the number that that function returned represents the total of items that are actually going to operate because both they and their ancestors are ALL "active" and how that is different than if it is not (and how previous Mudlet version prior to this change operated) because that (was/)is only checking the state of the identified item(s)...

Sorry, I don't follow. Thinking about it though the returned number of active items should not include ancestors - we are still looking at the items themselves, just changing the criteria of how an item is considered 'active', right?

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions

SlySven committed:
Revise: fixup for offset-timers that do not use mActive in the standard way

Also revise `isActive(...)` so that it optionally (with a third `true`
argument) returns `true` only if all the ancestors are also "active".

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions

vadi2 commented:
I've had a look at it but it still has isAncestorsActive(), isn't this now entirely covered by ancestors()?

SlySven commented:
> I've had a look at it but it still has isAncestorsActive(), isn't this now entirely covered by ancestors()?

`isAncestorsActive(...)` returns simple `true`/`false` for the item as to whether every parent is active or not - so provides the information that `isActive(...)` doesn't without the new option; `ancestors(...)` provides all the details that describe the successively ancient ancestors of the item - not just whether they are active or not but their name, ID and whether they are a package/module, a group/folder or an individual item (which can happen for trigger chains or offset timers IIRC).

Whilst I suppose the former is not essential it provides a simpler means to check the state of all the ancestors of an item rather than checking the `active` value of every member in the list type table that the latter produces...

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions

SlySven committed:
Fix: repair broken merge

The code merged in from the 'development` branch used an inlined getter
method that I had eliminated in this PR as it was working on a class public
member and it is simpler to use that member directly.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:books: Documentation being prepared in Area51 for the new functions:
* **[`ancestors(...)`](https://wiki.mudlet.org/w/Area_51#ancestors.2C_new_in_PR_.236726)**
* **[`isAncestorsActive(...)`](https://wiki.mudlet.org/w/Area_51#isAncestorsActive.2C_new_in_PR_.236726)**

Also revision to the existing function:
* **[`isActive(...)`](https://wiki.mudlet.org/w/Area_51#isActive.2C_modified_by_PR_.236726)**

SlySven committed:
Merge branch 'development' into Improve_add_ancestors_functions
